trilium_alchemy/core/note/note.py

```diff
@@ -936,17 +926,17 @@ def __ixor__(
         """
         parents = (
             {p for p in parent}
-            if isinstance(parent, Iterable) and not type(parent) is tuple
+            if isinstance(parent, Iterable) and type(parent) is not tuple
             else {parent}
         )
 
         self.branches.parents |= parents
 
         return self
 
-    def __getitem__(self, key: str | int) -> str | Note:
+    def __getitem__(self, key: str) -> str | Note:
         """
-        Return value of first attribute with provided name.
+        Return value of first attribute with provided name or index.

```

Comment:
The original comment was correct - the key can only be a name, not an index. Since when you access note["attributeName"] it returns the value of the attribute, not the attribute itself, it wouldn't really make sense to support accessing e.g. `note[0]` IMO. The user can always do `note.attributes[n]` if they want to access the nth attribute.

Reply:
Missed the comments when I fixed the typing hints

Reply:
Please find a stack trace from when I reintroduced `self._list = None`. I understand this is "self inflicted", should `__len()` add a model setup check?


```
INFO     root:session.py:161 Got Trilium version: 0.61.5-beta
___________________________________________________________ test_iter ____________________________________________________________

session = <trilium_alchemy.core.session.Session object at 0x1062a8310>
note = Note(title=Test note 2023-08-19 12:46:52.504629, note_id=bIjU7QHAp2Oy)

    @pytest.mark.attribute("Label5", "Value5")
    @pytest.mark.attribute("Label6", "Value6")
    @pytest.mark.attribute("Label7", "Value7")
    @pytest.mark.attribute("Label8", "Value8")
    @pytest.mark.attribute("Label9", "Value9")
    def test_iter(session: Session, note: Note):
        # len / iterate over attributes
>       assert len(note) == 5

note       = Note(title=Test note 2023-08-19 12:46:52.504629, note_id=bIjU7QHAp2Oy)
session    = <trilium_alchemy.core.session.Session object at 0x1062a8310>

test/note/test_note_attributes.py:49: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
trilium_alchemy/core/note/note.py:994: in __len__
    return len(self.attributes._name_map)
        self       = Note(title=Test note 2023-08-19 12:46:52.504629, note_id=bIjU7QHAp2Oy)
trilium_alchemy/core/entity/model.py:437: in _require_setup
    ent._model.setup_check()
        ent        = Note(title=Test note 2023-08-19 12:46:52.504629, note_id=bIjU7QHAp2Oy)
        func       = <function ExtensionDescriptor.__get__ at 0x104bc9bc0>
        objtype    = <class 'trilium_alchemy.core.note.note.Note'>
        self       = <trilium_alchemy.core.entity.model.ExtensionDescriptor object at 0x104bd5890>
trilium_alchemy/core/entity/model.py:177: in setup_check
    self.setup()
        self       = <trilium_alchemy.core.note.note.NoteModel object at 0x1062a9010>
trilium_alchemy/core/entity/model.py:255: in setup
    ext._setup(model_backing)
        create     = None
        ext        = <trilium_alchemy.core.note.attributes.OwnedAttributes object at 0x1062a8f90>
        model_backing = Note(note_id='bIjU7QHAp2Oy', title='Test note 2023-08-19 12:46:52.504629', type='text', mime='text/html', is_protected...23-08-19 19:46:52.577+0000', utc_date_created='2023-08-19 19:46:52.577Z', utc_date_modified='2023-08-19 19:46:52.577Z')
        self       = <trilium_alchemy.core.note.note.NoteModel object at 0x1062a9010>
trilium_alchemy/core/note/attributes.py:130: in _setup
    attribute.Attribute._from_model(
        attr_model = Attribute(attribute_id='5awaiyptBg7S', note_id='bIjU7QHAp2Oy', type='label', name='Label5', value='Value5', position=10, is_inheritable=False, utc_date_modified='2023-08-19 19:46:52.597Z')
        model      = Note(note_id='bIjU7QHAp2Oy', title='Test note 2023-08-19 12:46:52.504629', type='text', mime='text/html', is_protected...23-08-19 19:46:52.577+0000', utc_date_created='2023-08-19 19:46:52.577Z', utc_date_modified='2023-08-19 19:46:52.577Z')
        self       = <trilium_alchemy.core.note.attributes.OwnedAttributes object at 0x1062a8f90>
trilium_alchemy/core/attribute/attribute.py:214: in _from_model
    attr = label.Label(
        cls        = <class 'trilium_alchemy.core.attribute.attribute.Attribute'>
        label      = <module 'trilium_alchemy.core.attribute.label' from '/Users/honce/Projects/Python/trilium-alchemy/trilium_alchemy/core/attribute/label.py'>
        model      = Attribute(attribute_id='5awaiyptBg7S', note_id='bIjU7QHAp2Oy', type='label', name='Label5', value='Value5', position=10, is_inheritable=False, utc_date_modified='2023-08-19 19:46:52.597Z')
        owning_note = Note(title=Test note 2023-08-19 12:46:52.504629, note_id=bIjU7QHAp2Oy)
        relation   = <module 'trilium_alchemy.core.attribute.relation' from '/Users/honce/Projects/Python/trilium-alchemy/trilium_alchemy/core/attribute/relation.py'>
        session    = <trilium_alchemy.core.session.Session object at 0x1062a8310>
trilium_alchemy/core/session.py:632: in wrapper
    return func(*args, **kwargs)
        args       = (Label(#Label5, value=Value5, attribute_id=5awaiyptBg7S, note=Note(title=Test note 2023-08-19 12:46:52.504629, note_id=bIjU7QHAp2Oy), position=10), 'Label5')
        func       = <function Label.__init__ at 0x104bd0540>
        kwargs     = {'attribute_id': '5awaiyptBg7S', 'model_backing': Attribute(attribute_id='5awaiyptBg7S', note_id='bIjU7QHAp2Oy', type=...-08-19 12:46:52.504629, note_id=bIjU7QHAp2Oy), 'session': <trilium_alchemy.core.session.Session object at 0x1062a8310>}
trilium_alchemy/core/entity/model.py:21: in wrapper
    return func(*args, **kwargs)
        args       = (Label(#Label5, value=Value5, attribute_id=5awaiyptBg7S, note=Note(title=Test note 2023-08-19 12:46:52.504629, note_id=bIjU7QHAp2Oy), position=10), 'Label5')
        func       = <function Label.__init__ at 0x104bd04a0>
        kwargs     = {'attribute_id': '5awaiyptBg7S', 'model_backing': Attribute(attribute_id='5awaiyptBg7S', note_id='bIjU7QHAp2Oy', type=...-08-19 12:46:52.504629, note_id=bIjU7QHAp2Oy), 'session': <trilium_alchemy.core.session.Session object at 0x1062a8310>}
trilium_alchemy/core/attribute/label.py:50: in __init__
    super().__init__(
        __class__  = <class 'trilium_alchemy.core.attribute.label.Label'>
        inheritable = False
        kwargs     = {'attribute_id': '5awaiyptBg7S', 'model_backing': Attribute(attribute_id='5awaiyptBg7S', note_id='bIjU7QHAp2Oy', type=...ied='2023-08-19 19:46:52.597Z'), 'owning_note': Note(title=Test note 2023-08-19 12:46:52.504629, note_id=bIjU7QHAp2Oy)}
        model_backing = Attribute(attribute_id='5awaiyptBg7S', note_id='bIjU7QHAp2Oy', type='label', name='Label5', value='Value5', position=10, is_inheritable=False, utc_date_modified='2023-08-19 19:46:52.597Z')
        name       = 'Label5'
        self       = Label(#Label5, value=Value5, attribute_id=5awaiyptBg7S, note=Note(title=Test note 2023-08-19 12:46:52.504629, note_id=bIjU7QHAp2Oy), position=10)
        session    = <trilium_alchemy.core.session.Session object at 0x1062a8310>
        value      = ''
trilium_alchemy/core/session.py:632: in wrapper
    return func(*args, **kwargs)
        args       = (Label(#Label5, value=Value5, attribute_id=5awaiyptBg7S, note=Note(title=Test note 2023-08-19 12:46:52.504629, note_id=bIjU7QHAp2Oy), position=10), 'Label5')
        func       = <function Attribute.__init__ at 0x104bcbb00>
        kwargs     = {'attribute_id': '5awaiyptBg7S', 'inheritable': False, 'model_backing': Attribute(attribute_id='5awaiyptBg7S', note_id...2023-08-19 19:46:52.597Z'), 'owning_note': Note(title=Test note 2023-08-19 12:46:52.504629, note_id=bIjU7QHAp2Oy), ...}
trilium_alchemy/core/entity/model.py:21: in wrapper
    return func(*args, **kwargs)
        args       = (Label(#Label5, value=Value5, attribute_id=5awaiyptBg7S, note=Note(title=Test note 2023-08-19 12:46:52.504629, note_id=bIjU7QHAp2Oy), position=10), 'Label5')
        func       = <function Attribute.__init__ at 0x104bcba60>
        kwargs     = {'attribute_id': '5awaiyptBg7S', 'inheritable': False, 'model_backing': Attribute(attribute_id='5awaiyptBg7S', note_id...2023-08-19 19:46:52.597Z'), 'owning_note': Note(title=Test note 2023-08-19 12:46:52.504629, note_id=bIjU7QHAp2Oy), ...}
trilium_alchemy/core/attribute/attribute.py:62: in _require_attribute_id
    return func(ent, *args, **kwargs)
        args       = ('Label5',)
        ent        = Label(#Label5, value=Value5, attribute_id=5awaiyptBg7S, note=Note(title=Test note 2023-08-19 12:46:52.504629, note_id=bIjU7QHAp2Oy), position=10)
        func       = <function Attribute.__init__ at 0x104bcb9c0>
        kwargs     = {'attribute_id': '5awaiyptBg7S', 'inheritable': False, 'model_backing': Attribute(attribute_id='5awaiyptBg7S', note_id...2023-08-19 19:46:52.597Z'), 'owning_note': Note(title=Test note 2023-08-19 12:46:52.504629, note_id=bIjU7QHAp2Oy), ...}
trilium_alchemy/core/attribute/attribute.py:173: in __init__
    if owning_note:
        __class__  = <class 'trilium_alchemy.core.attribute.attribute.Attribute'>
        attribute_id = '5awaiyptBg7S'
        inheritable = False
        kwargs     = {}
        model_backing = Attribute(attribute_id='5awaiyptBg7S', note_id='bIjU7QHAp2Oy', type='label', name='Label5', value='Value5', position=10, is_inheritable=False, utc_date_modified='2023-08-19 19:46:52.597Z')
        name       = 'Label5'
        owning_note = Note(title=Test note 2023-08-19 12:46:52.504629, note_id=bIjU7QHAp2Oy)
        self       = Label(#Label5, value=Value5, attribute_id=5awaiyptBg7S, note=Note(title=Test note 2023-08-19 12:46:52.504629, note_id=bIjU7QHAp2Oy), position=10)
        session    = <trilium_alchemy.core.session.Session object at 0x1062a8310>
trilium_alchemy/core/note/note.py:994: in __len__
    return len(self.attributes._name_map)
        self       = Note(title=Test note 2023-08-19 12:46:52.504629, note_id=bIjU7QHAp2Oy)
trilium_alchemy/core/note/attributes.py:58: in _name_map
    for attr in list(self):
        attrs      = {}
        self       = <trilium_alchemy.core.note.attributes.Attributes object at 0x1062a9210>
trilium_alchemy/core/note/attributes.py:422: in __iter__
    yield from self._combined
        self       = <trilium_alchemy.core.note.attributes.Attributes object at 0x1062a9210>
trilium_alchemy/core/note/attributes.py:406: in _combined
    return list(self._owned) + list(self._inherited)
        self       = <trilium_alchemy.core.note.attributes.Attributes object at 0x1062a9210>
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <trilium_alchemy.core.note.attributes.InheritedAttributes object at 0x106266250>

    def __iter__(self):
>       yield from self._list
E       TypeError: 'NoneType' object is not iterable

self       = <trilium_alchemy.core.note.attributes.InheritedAttributes object at 0x106266250>
```

Reply:
Ah, this is a subtle one!

Subclassing `MutableMapping` does introduce a pitfall: Python changed the way it evaluates `Note` for truthiness. Now it will be based on whether or not it has any attributes, as it's invoking `__len__()` as per the stack trace.

So on this line:

```
trilium_alchemy/core/attribute/attribute.py:173: in __init__
    if owning_note:
```

The check should actually be:

```python
    if owning_note is not None:
```

Here `owning_note` is the one which is currently being setup, so its inherited attributes weren't setup yet. Hopefully this is the only such place in the code - I think I explicitly used `is not None` most places.

Reply:
done